aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,163 @@
+
+  int64_t linear_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_index >= N) {
+    return;
+  }
+
+  linear_index_to_indices(linear_index, each_dim_len, total_dims, indices);
+
+  // flip nD index along each desired dimension
+  for (int64_t i = 0 ; i < flip_dims_size; i++) {
+    int64_t dim = dims[i];
+    int64_t indices_dim = linear_index * total_dims + dim;
+    indices[indices_dim] = shape[dim] - 1 - indices[indices_dim];
+  }
+  int64_t dest_linear_index = indices_to_linear_index(indices, total_dims, each_dim_len, linear_index);
+  out_t[linear_index] = in_t[dest_linear_index];
+}
+
+Tensor flip_cuda(const Tensor& self, IntList dims) {
+
+  // TODO: allow non-contiguous tensors
+  self.contiguous();
+
+  int64_t flip_dims_size = dims.size(), total_dims = self.dim(), N = self.numel();
+
+  // check if number of axis in dim is valid
+  if (flip_dims_size == 0) {
+    std::stringstream ss;
+    ss << "expected input tensor dims not empty, "
+       << "but got tensor dims size=" << flip_dims_size;
+    throw std::runtime_error(ss.str());
+  }
+
+  // check duplicates in dims
+  auto flip_dims_v = std::vector<int64_t>(dims);
+  flip_dims_v.erase(std::unique(flip_dims_v.begin(), flip_dims_v.end()), flip_dims_v.end());
+  if ((int64_t)flip_dims_v.size() < flip_dims_size) {
+    std::stringstream ss;
+    ss << "dims has duplicates, "
+       << "original flip dims size=" << flip_dims_size << ", "
+       << "but unique flip dims size= " << flip_dims_v.size();
+    throw std::runtime_error(ss.str());
+  }
+
+  // check len of dims
+  if (flip_dims_size > total_dims) {
+    std::stringstream ss;
+    ss << "expected flip dims size <= tensor total dims, "
+       << "but got flip dims size=" << flip_dims_size << " and "
+       << "tensor total dim=" << total_dims;
+    throw std::runtime_error(ss.str());
+  }
+
+  // check if dims axis within range
+  int64_t min_d = total_dims, max_d = 0;
+  for (auto d : dims) {
+    min_d = std::min(min_d, d);
+    max_d = std::max(max_d, d);
+  }
+
+  if (min_d < 0) {
+    std::stringstream ss;
+    ss << "expected flip dims axis >= 0, "
+       << "but got min flip dims=" << min_d;
+    throw std::runtime_error(ss.str());
+  }
+
+  if (max_d >= total_dims) {
+    std::stringstream ss;
+    ss << "expected flip dims axis < tensor total dims, "
+       << "but got max flip dims=" << max_d << " and "
+       << "tensor total dim=" << total_dims;
+    throw std::runtime_error(ss.str());
+  }
+
+  Tensor flip_dims_t = at::zeros(CPU(kLong), {flip_dims_size});
+  int64_t* flip_dims_t_d = flip_dims_t.data<int64_t>();
+  for (int64_t i = 0; i < flip_dims_size; i++) {
+    flip_dims_t_d[i] = dims[i];
+  }
+
+  auto shape = self.sizes();
+  Tensor shape_t = at::zeros(CPU(kLong), {total_dims});
+  int64_t* shape_t_d = shape_t.data<int64_t>();
+  for (int64_t i = 0; i < total_dims; i++) {
+    shape_t_d[i] = shape[i];
+  }
+
+  Tensor each_dim_len = at::zeros(CPU(kLong), {total_dims});
```

Comment:
What is each_dim_length and how does it differ from the sizes of a tensor?

Reply:
I am thinking of tensor as a nD array. Each dimension has a length equals to the number of subarray at that dimension. For instance, let t = [[1,2], [3,4], [5,6]], then each_dim_length[0] = 3, each_dim_length[1] = 2

Reply:
so.... it is just size?

Reply:
@SsnL you are right, each_dim_len is the same as strides!